components/layout_2020/display_list/stacking_context.rs

```diff
@@ -0,0 +1,369 @@
+                stacking_context.fragments.push(StackingContextFragment {
+                    space_and_clip: builder.current_space_and_clip,
+                    containing_block: *containing_block,
+                    fragment: self,
+                });
+            },
+        }
+    }
+}
+
+impl BoxFragment {
+    fn get_stacking_context_type(&self) -> Option<StackingContextType> {
+        if self.establishes_stacking_context() {
+            return Some(StackingContextType::Real);
+        }
+
+        if self.style.get_box().position != ComputedPosition::Static {
+            return Some(StackingContextType::PseudoPositioned);
+        }
+
+        if self.style.get_box().float != ComputedFloat::None {
+            return Some(StackingContextType::PseudoFloat);
+        }
+
+        None
+    }
+
+    /// Returns true if this fragment establishes a new stacking context and false otherwise.
+    fn establishes_stacking_context(&self) -> bool {
+        if self.style.get_effects().opacity != 1.0 {
+            return true;
+        }
+
+        if self.style.get_effects().mix_blend_mode != ComputedMixBlendMode::Normal {
+            return true;
+        }
+
+        if self.has_filter_transform_or_perspective() {
+            return true;
+        }
+
+        if self.style.get_box().transform_style == ComputedTransformStyle::Preserve3d ||
+            self.style.overrides_transform_style()
+        {
+            return true;
+        }
+
+        // Fixed position and sticky position always create stacking contexts.
+        // TODO(mrobinson): We need to handle sticky positioning here when we support it.
+        if self.style.get_box().position == ComputedPosition::Fixed {
+            return true;
+        }
+
+        // Statically positioned fragments don't establish stacking contexts if the previous
+        // conditions are not fulfilled. Furthermore, z-index doesn't apply to statically
+        // positioned fragments.
+        if self.style.get_box().position == ComputedPosition::Static {
+            return false;
+        }
+
+        // For absolutely and relatively positioned fragments we only establish a stacking
+        // context if there is a z-index set.
+        // See https://www.w3.org/TR/CSS2/visuren.html#z-index
+        !self.style.get_position().z_index.is_auto()
+    }
+
+    // Get the effective z-index of this fragment. Z-indices only apply to positioned element
+    // per CSS 2 9.9.1 (http://www.w3.org/TR/CSS2/visuren.html#z-index), so this value may differ
+    // from the value specified in the style.
+    fn effective_z_index(&self) -> i32 {
+        match self.style.get_box().position {
+            ComputedPosition::Static => {},
+            _ => return self.style.get_position().z_index.integer_or(0),
+        }
+
+        if !self.style.get_box().transform.0.is_empty() {
+            return self.style.get_position().z_index.integer_or(0);
+        }
```

Comment:
My reading of https://drafts.csswg.org/css-transforms/#transform-rendering is that this should be removed:

> For elements whose layout is governed by the CSS box model, any value other than `none` for the `transform` property results in the creation of a stacking context. Implementations must paint the layer it creates, within its parent stacking context, at the same stacking order that would be used if it were a positioned element with `z-index: 0`. If an element with a transform is positioned, the `z-index` property applies as described in [CSS2], except that `auto` is treated as 0 since a new stacking context is always created.

Reply:
This is quite odd, because Gecko has a strange behavior here that Servo has been replicating. Observe how this particular file is rendered in Gecko:
```
<div style="position: absolute; top: 2px; z-index: 2; background: green; width: 100px; height: 100px;"></div>
<div style="background: blue; transform: rotate(45deg); z-index: 3; width: 100px; height: 100px;"></div>
```
In Gecko, the z-index of the transformed div is taken into account, while in Chromium and WebKit it is not.

Reply:
This is arguably a Gecko bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1256980

Another bit of relevant spec is https://drafts.csswg.org/css2/visuren.html#z-index with “Applies to: positioned elements”

Reply:
Oh, thanks for finding that bug. I've pushed a new version of the branch which removes this, bringing layout_2020 in line with WebKit and Blink.